### PR TITLE
Fix audit revision query for enum column mapping

### PR DESCRIPTION
**Pull request by X-Coder264**
## Subject

I am targeting this branch, because it's a bug fix.

After https://github.com/doctrine/orm/pull/10088 we get an enum object back in the `DEL` revision `saveRevisionEntityData` code path instead of an integer or string value which then blows up on the `$conn->executeStatement()` call with the following exception:

```
Error: Object of class Sonata\EntityAuditBundle\Tests\Fixtures\Issue\Status could not be converted to string

vendor/doctrine/dbal/src/Driver/PDO/Statement.php:56
```

## Changelog

```markdown
### Fixed
- Audit query for Doctrine ORM >= 2.14.1 for entities with enumType column mapping
```

### CODE REVIEW

**Review by VincentLanglet — CHANGES_REQUESTED** [review]
Thanks for the fix, do you have time to take a look at the ci errors ?
You'll need to
- Solve some linters errors
- Solve some "possiblyNull" static analysis error by adding `static::assertNotNull($foo)`.
- Skip the test for PHP version which doesn't support ENUM

File: src/EventListener/LogRevisionsListener.php [hunk 1]
```diff
@@ -582,6 +582,12 @@ private function saveRevisionEntityData(EntityManagerInterface $em, ClassMetadat
             );
         }
 
+        foreach ($params as $key => $parameterValue) {
+            if (interface_exists(\BackedEnum::class) && $parameterValue instanceof \BackedEnum) {
```
Comment (phansys) [inline src/EventListener/LogRevisionsListener.php]: Do we really need the `interface_exists()` check?

File: composer.json [hunk 2]
```diff
@@ -14,7 +14,7 @@
         "doctrine/common": "^3.1",
         "doctrine/dbal": "^2.13 || ^3.0",
         "doctrine/event-manager": "^1.1",
-        "doctrine/orm": "^2.10",
+        "doctrine/orm": "^2.11",
```
Comment (phansys) [inline composer.json]: IMO, this change is not related to the described fix.

File: src/EventListener/LogRevisionsListener.php (hunk 1 above)
Comment (X-Coder264) [reply src/EventListener/LogRevisionsListener.php]: I don't know how PHP behaves when doing an instanceof check against an interface that does not exist, even if that maybe doesn't emit a warning today it doesn't mean it's not going to in the future so I think it's better to keep this as is as it definitively doesn't hurt.

File: composer.json (hunk 2 above)
Comment (X-Coder264) [reply composer.json]: 2.11 added the `enumType` feature for the mapping which is needed to reproduce the issue (and is used in the test) but I see now that you test PHP 8.1 only against `highest` dependencies so I should be able to revert this.

**Comment by X-Coder264** [comment]
Even though the test I've added is marked as requiring PHP 8.1 to run, the PHP 7.4 and PHP 8.0 pipelines are still failing as the entity with the enum mapping class metadata is still being registered (so it blows up in `Doctrine/ORM/Mapping/ClassMetadataFactory`). Does anybody have a suggestion how we can solve this?

**Comment by VincentLanglet** [comment]
> Even though the test I've added is marked as requiring PHP 8.1 to run, the PHP 7.4 and PHP 8.0 pipelines are still failing as the entity with the enum mapping class metadata is still being registered (so it blows up in `Doctrine/ORM/Mapping/ClassMetadataFactory`). Does anybody have a suggestion how we can solve this?

Filtering here maybe https://github.com/sonata-project/EntityAuditBundle/blob/1.x/tests/BaseTest.php#L98-L102

File: src/EventListener/LogRevisionsListener.php (hunk 1 above)
Comment (phansys) [reply src/EventListener/LogRevisionsListener.php]: PHP handles the absence of the class gracefully. Avoiding superfluous function calls is always desirable.
```suggestion
            if ($parameterValue instanceof \BackedEnum) {
```

Comment (X-Coder264) [reply src/EventListener/LogRevisionsListener.php]: Done

File: composer.json (hunk 2 above)
Comment (X-Coder264) [reply composer.json]: Done

**Comment by X-Coder264** [comment]
> > Even though the test I've added is marked as requiring PHP 8.1 to run, the PHP 7.4 and PHP 8.0 pipelines are still failing as the entity with the enum mapping class metadata is still being registered (so it blows up in `Doctrine/ORM/Mapping/ClassMetadataFactory`). Does anybody have a suggestion how we can solve this?
> 
> Filtering here maybe https://github.com/sonata-project/EntityAuditBundle/blob/1.x/tests/BaseTest.php#L98-L102

@VincentLanglet Thanks - all CI pipelines have now passed :tada:

**Review by VincentLanglet — APPROVED** [review]

**Review by phansys — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks
